files_orpg_sw/man/cat1/install_adapt.1
install_adapt(1)		Open RPG		install_adapt(1)

NAME
    install_adapt - Unpacks an RPG adaptation data archive file and installs 
    the files in the local RPG.

SYNOPSIS
   install_adapt [options]

DESCRIPTION
    Searches in a specified directory to find an RPG adaptation data archive
    file, unpacks it and installs the unpacked files in the local RPG. This
    routine is called as part of mrpg startup to install any newly restored
    adaptation data archives. RPG must be shutdown to install adaptation data.
    Execution of this program in transition mode is allowed. install_adapt 
    must be invoked on a RPG host. It only installs the archive that matches
    the radar site, RPG adapt version number and the RPG redundant channel.

    The "-s" option is used to replace the "find_adapt" tool. It is expected 
    to be used only by build_adapt_floppy. With this option, install_adapt does
    not performs installation. It, instead, prints the RPG archive file names
    of all redundant channels that matches "site" as well as "date" and "time" 
    as specified.

Options
    -D dir (Specifies the directory where RPG adapt archive files 
       to be found. Default: $CFG_DIR/adapt.)
    -d date (mm/dd/yyyy - Archive file date to install. Default: Today)
    -t time (hh:mm:ss - Archive file time to install. Default: The 
       current time.)
       Note: If date is specified but not time, the latest file of the 
       date is the best match. Otherwise, the file closest to the 
       specified time is the best match.
    -m (Moves adaptation archive files to installed and not installed 
       sub dirs after installation.)
    -x text (Prints "text" on the screen when installing.)
    -s site (Prints adaptation archive file names for all redundant 
       channels for site \"site\" to support build_adapt_floppy. 
       Installation is not performed.)
    -h (Print usage info)

EXAMPLES

    Searches adaptation data archive of date "03/03/2005" in the current 
    directory and install it. The latest archive of the specified date is 
    installed.

    install_adapt -D . -d 03/03/2005

    Searches adaptation data archive of local time "23:00:00" of date 
    "03/03/2005" in directory "$CFG_DIR/adapt" and install it. The archive 
    closest to the specified time is installed.

    install_adapt -d 03/03/2005 -t "23:00:00"

    Install the latest archive for this site that is located in the
    $CFG_DIR/adapt directory. Once the archive is installed, move the archive
    to the $CFG_DIR/adapt/installed directory. Any archives that existed in
    $CFG_DIR/adapt that were not installed will be moved to the
    $CFG_DIR/adapt/not_installed directory. Echo the message "Installing
    Factory Defaults" during the installation of the archive. Use "lem
    install_adapt" to see the files that are installed.

    install_adapt -D $CFG_DIR/adapt -m -x "Install Factory Defaults"

INPUTS
    orpg adaptation archive

SEE ALSO
    save_adapt.1, restore_adapt.1, save_adapt_floppy.1, restore_adapt_floppy.1


files_orpg_sw/src/cpc026/tsk001/install_adapt.c

/******************************************************************

    This is the main module for install_adapt.
	
******************************************************************/

/* 
 * RCS info
 * $Author: jing $
 * $Locker:  $
 * $Date: 2005/03/21 15:59:32 $
 * $Id: install_adapt.c,v 1.13 2005/03/21 15:59:32 jing Exp $
 * $Revision: 1.13 $
 * $State: Exp $
 */  

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <string.h>

#include <orpg.h>
#include <orpgadpt.h>
#include <infr.h>

/* command line options */
static char *Source_dir = "";
static char *Source_date = "";
static char *Source_time = "";
static char *Text_out = "";
static int Need_file_move = 0;

static int Read_options (int argc, char **argv);
static void Print_usage (char **argv);


/******************************************************************

    The main function.

******************************************************************/

int main (int argc, char **argv) {
    int ret;

    if (Read_options (argc, argv) != 0)
	exit (1);

    if ((ret = LE_set_option ("LB type", LB_SINGLE_WRITER)) < 0 || 
	(ret = LE_init (argc, argv)) < 0) {
	LE_send_msg (0, "LE_init failed (%d)", ret);
	exit (1);
    }

    if (strlen (Source_dir) == 0) {	/* The default source dir */
	if ((Source_dir = malloc (ADPTU_NAME_SIZE)) == NULL) {
	    LE_send_msg (0, "malloc failed");
	    exit (1);
	}
	ret = MISC_get_cfg_dir (Source_dir, ADPTU_NAME_SIZE - 8);
	if (ret < 0) {
	    LE_send_msg (0, "MISC_get_cfg_dir failed (%d)", ret);
	    exit (1);
	}
	strcat (Source_dir, "/adapt");
    }

    if (strlen (Text_out) > 0)
	printf ("%s\n", Text_out);
    ORPGADPTU_install_adapt (Source_dir, Source_date, 
				Source_time, Need_file_move);

    exit (0);
}

/**************************************************************************

    Description: This function reads command line arguments.

    Inputs:	argc - number of command arguments
		argv - the list of command arguments

    Return:	It returns 0 on success or -1 on failure.

**************************************************************************/

static int Read_options (int argc, char **argv) {
    extern char *optarg;    /* used by getopt */
    extern int optind;
    int c;                  /* used by getopt */
    int err;                /* error flag */

    err = 0;
    while ((c = getopt (argc, argv, "D:d:t:s:mOx:qh?")) != EOF) {
	switch (c) {

	    case 'D':
		Source_dir = optarg;
		break;

	    case 'd':
		Source_date = optarg;
		break;

	    case 't':
		Source_time = optarg;
		break;

	    case 'x':
		Text_out = optarg;
		break;

	    case 'm':
		Need_file_move = 1;
		break;

	    case 's':
	    case 'O':
	    case 'q':
		break;

	    case 'h':
	    case '?':
		Print_usage (argv);
		break;
	}
    }

    return (err);
}

/**************************************************************************

    Description: This function prints the usage info.

**************************************************************************/

static void Print_usage (char **argv) {
    char *usage = "\
    Searches an RPG adaptation data archive in a specified directory,\n\
    which must be local, and installs it in the current RPG environment.\n\
    The file that best matches the specified date and time is used.\n\
    Adaptation data is only installed on the RPG nodes. RPG adaptation\n\
    version number and site name (e.g. KTLX) must match.\n\
    Options:\n\
	-D dir (Specifies the directory where RPG adapt archive files \n\
	   to be found. Default: $CFG_DIR/adapt.)\n\
	-d date (mm/dd/yyyy - Archive file date to install. Default: Today)\n\
	-t time (hh:mm:ss - Archive file time to install. Default: The \n\
	   current time.)\n\
	   Note: If date is specified but not time, the latest file of the \n\
	   date is the best match. Otherwise, the file closest to the \n\
	   specified time is the best match.\n\
	-m (Moves adaptation archive files to installed and not installed \n\
	   sub dirs after installation.)\n\
	-x text (Prints \"text\" on the screen when installing.)\n\
	-h (Print usage info)\n\
";

    printf ("Usage:  %s [options]\n", argv[0]);
    printf ("%s\n", usage);

    exit (0);
}


